### ISSUE TITLE
Structured logging via StdOut/StdErr does not seem to show up in stackdriver for App Engine Flexible

### ISSUE THREAD

**Issue by mahesh-bennie** [GoogleCloudPlatform/fluent-plugin-google-cloud#401]
I have a dotnet core (3.1) api running on a GAE Flex VM. I have been following the samples in this documentation (https://cloud.google.com/run/docs/logging#writing_structured_logs) and this is what I had come up with:
```
private void SendLog(string message, LogLevel logLevel)
        {
            var log = new StructuredLog
            {
                Category = _category,
                Severity = logLevel.ToLogSeverity(),
                Message = message,
            };

            if (_accessor?.HttpContext != null)
            {
                var traceId = _accessor.HttpContext.Request.Headers["X-Cloud-Trace-Context"].ToString();
                if (!string.IsNullOrWhiteSpace(traceId))
                {
                    log.Trace = $"projects/{_configuration["GCP_PROJ_NAME"]}/traces/{traceId}";
                }
                log.User = _accessor.HttpContext.User.Claims.SingleOrDefault(x => x.Type == "UserId")?.ToString();
            }

            var serializedLog = JsonConvert.SerializeObject(log);
            Console.WriteLine(serializedLog);
        }
```
Where StructuredLog is a simple POCO:
```
internal class StructuredLog
        {
            [JsonProperty("user")]
            public string User { get; set; }
            [JsonProperty("category")]
            public string Category { get; set; }
            [JsonProperty("severity")]
            public LogSeverity Severity { get; set; }
            [JsonProperty("message")]
            public string Message { get; set; }
            [JsonProperty("logging.googleapis.com/trace")]
            public string Trace { get; set; }
       }
```
I then hook this class (that contains said method) as an ILogger and via loggerfactory at startup use it in various place via DI like so:
```
_logger.LogInformation("Begin logging...");
```
An example of how this message would look like:
```
{"category":"Example.Api.Middleware.LoggingMiddleware","severity":200,"message":"Begin logging..."}
```
I just don't see anything show up under StackDriver. I looked at every logtype available (stdout, stderr, nginx request, etc) but it never shows up. 
A simple Console.WriteLine("Some message") does appear as textpayload but not this structured log. 
Unclear what else could possibly be needed to get this working as according to documentation, everything should already have been setup for Google App Engine Flexible out of the box.

**Comment by igorpeshansky** [comment]
As you've discovered, logs in App Engine Flexible are handled by the Stackdriver logging agent, and the structure it expects is documented in https://cloud.google.com/logging/docs/agent/configuration#process-payload. If text logs end up in Cloud Logging, but structured logs don't, you might be able to see some messages in the logs from the logging agent container. @yanweiguo might be able to offer more insight there.

For testing purposes, you could use `Console.WriteLine` to write a JSON string directly, just to make sure the format is what you expect. If those messages do appear in Cloud Logging, then you should debug your `SendLog` method. If they don't, we can continue debugging that (much simpler) use case.

One possible workaround would be to use the direct Cloud Logging API integration (https://cloud.google.com/logging/docs/integrate/dotnet).

**Comment by yanweiguo** [comment]
cc @Sanrag

**Comment by mahesh-bennie** [comment]
@igorpeshansky hey thanks for responding. That JsonConvert.SerializeObject just translates the object to a string (I literally copied and pasted that example payload while debugging). I tried using Console.Write after reading one of your comments https://github.com/GoogleCloudPlatform/appengine-sidecars-docker/pull/81#issuecomment-364108839 
as WriteLine adds a line terminator but still didn't work. 

I went ahead and took a look at the logging agent like you mentioned and saw a ton of these messages appear (which seem to correlate to when a log was supposed to have been fired)
```
Unable to export to Monitoring service because: GaxError RPC failed, caused by 3:One or more TimeSeries could not be written: Metrics cannot be written to gae_app. See https://cloud.google.com/monitoring/custom-metrics/creating-metrics#which-resource for a list of writable resource types.: timeSeries[0]
```
I did some digging on that and came to this SO but it wasn't really forthcoming on what's going on. It does allude to the logs but doesn't give out a solution.
https://stackoverflow.com/questions/59738710/unable-to-export-to-monitering-service-because-gaxerror-rpc-failed-caused-by-3 
Could this be the issue? If so, what does it mean and is there something I can do on my end to get it working?

**Comment by mahesh-bennie** [comment]
And I just read deeper into that SO responses and saw @yanweiguo respond to it a while back. Does this mean there's currently a showstopper on GAE Flex that prevents us from using StdOut/StdErr? Only reason we're hesitant to using the client library is because it introduces lag and we really want our traces to be as accurate as possible with regards to latency and not having to worry about extraneous API calls.

**Comment by yanweiguo** [comment]
Hi @mahesh-bennie, when did you deploy your app? The `Unable to export` logs issue should be fixed months ago.

Flex flex supports structure logs. I just deployed a Python app which sends the same log to stdout:
```
{"category":"Example.Api.Middleware.LoggingMiddleware","severity":200,"message":"Begin logging..."}
```

The `Begin logging...` part showed up in cloud logging.

I'd suggest ssh into the Flex VM instance and check what the logs look like which your app send to stdout/stderr. You can use `docker logs gaeapp` or find the logs under `/var/log/saved_docker`.

**Comment by mahesh-bennie** [comment]
@yanweiguo we deploy the app almost everyday across our environments (dev,stage,prod)
Yeah the documentation clearly indicates as such as well, which is why I'm confused why it's not showing up. 
I did ssh into the instance as you suggested and after running the command, the only logs I see are the unstructured logs (simple one liner strings that show up as textpayload on the Stackdriver UI).

**Comment by yanweiguo** [comment]
Then the problem is that somehow the structured log was not sent to stdout. Did you see it from `docker logs` command? If not, I'd suggest debugging the way how you send logs. (I'm not familiar with dotnet)

**Comment by mahesh-bennie** [comment]
Yup that definitely was the case. There was a configuration that disables any loglevels under warning from being sent over the streams. Figured it out after testing out with different log levels. 
tl;dr; bad config issue. 
Thank you guys for your patience and guidance! Closing this now.
